src/System.IO.Ports/src/System/IO/Ports/SerialStream.Unix.cs

```diff
@@ -563,27 +624,231 @@ internal SafeFileHandle OpenPort(string portName)
+                }
+                // ignore EWOULDBLOCK since we handle timeout elsewhere
+                else if (lastError != Interop.Error.EWOULDBLOCK)
+                {
+                    r.Complete(new IOException());
+                }
+            }
+
+            return 0;
+        }
+
+        private unsafe int ProcessWrite(SerialStreamAsyncResult r)
+        {
+            ReadOnlySpan<byte> buff = r.Buffer.Span;
+            fixed (byte* bufPtr = buff)
+            {
+                // assumes dequeue-ing happens on a single thread
+                int numBytes = Interop.Sys.Write(_handle, bufPtr, buff.Length);
+
+                Interop.Error lastError = numBytes < 0 ? Interop.Sys.GetLastError() : 0;
+                if (numBytes > 0)
+                {
+                    r.ProcessBytes(numBytes);
+
+                    if (r.Buffer.Length == 0)
                     {
+                        r.Complete();
                     }
-                    base.Dispose(disposing);
+
+                    return numBytes;
+                }
+                // ignore EWOULDBLOCK since we handle timeout elsewhere
+                // numBytes == 0 means that there might be an error
+                else if (lastError != Interop.Error.SUCCESS && lastError != Interop.Error.EWOULDBLOCK)
+                {
+                    r.Complete(new IOException());
+                }
+            }
+
+            return 0;
+        }
+
+        // returns number of bytes read/written
+        private static int DoIORequest(ConcurrentQueue<SerialStreamAsyncResult> q, RequestProcessor op)
+        {
+            // assumes dequeue-ing happens on a single thread
+            while (q.TryPeek(out SerialStreamAsyncResult r))
+            {
+                if (r.IsCompleted)
+                {
+                    Debug.Assert(q.TryDequeue(out _));
+                    // take another item since we haven't processed anything
+                    continue;
+                }
+
+                int ret = op(r);
+                Debug.Assert(ret >= 0);
+
+                if (r.IsCompleted)
+                {
+                    Debug.Assert(q.TryDequeue(out _));
+                }
+
+                return ret;
+            }
+
+            return 0;
+        }
+
+        private unsafe void IOLoop()
+        {
+            bool eofReceived = false;
+            //bool readyForReceivedEvent = true;
+            long totalBytesRead = 0;
+            // last value of totalBytesRead + BytesToRead
+            long lastTotalBytesAvailable = BytesToRead;
+            while (IsOpen && !eofReceived && !_ioLoopFinished)
+            {
+                // PollEvents will wait if both flags are false
+                Interop.Sys.PollEvents events = PollEvents(1,
+                                                           pollReadEvents: _readQueue.Count > 0,
+                                                           pollWriteEvents: _writeQueue.Count > 0);
```

Comment:
I've changed the design of how this works a bit since last time you looked at it - currently the DataReceived works independently of poll's result. New design allows to rise events multiple times even when driver's data is being filled and not only when reading happens and new data shows up.
This is because I previously didn't notice that this event is only used by the logic on top of this event and the event with the same name is also on SerialPort class which adds more logic on top of it (i.e. allows adding number of bytes which needs to be available before event is raised)